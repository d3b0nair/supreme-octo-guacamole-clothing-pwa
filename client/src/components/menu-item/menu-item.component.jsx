import React from "react";
import { withRouter } from "react-router-dom";

import {
  MenuItemContainer,
  BackgroundImageContainer,
  ContentContainer,
  ContentTitle,
  ContentSubtitle,
} from "./menu-item.styles";

export const MenuItem = ({
  title,
  imageUrl,
  size,
  history,
  linkUrl,
  match,
}) => (
  <MenuItemContainer
    data-testid="MenuItemContainer"
    size={size}
    onClick={() => history.push(`${match.url}${linkUrl}`)}
  >
    <BackgroundImageContainer
      className="background-image"
      imageUrl={`${window.location.origin}/${imageUrl}`}
    />
    <ContentContainer className="content">
      <ContentTitle>{title.toUpperCase()}</ContentTitle>
      <ContentSubtitle>SHOP NOW</ContentSubtitle>
    </ContentContainer>
  </MenuItemContainer>
);

export default withRouter(MenuItem);
