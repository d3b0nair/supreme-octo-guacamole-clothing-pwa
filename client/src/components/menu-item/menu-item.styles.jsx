import styled from "styled-components";

export const MenuItemContainer = styled.div`
  height: ${({ size }) => (size ? "380px" : "240px")};
  min-width: 30%;
  overflow: hidden;
  flex: 1 1 auto;
  display: flex;
  align-items: center;
  justify-content: center;
  margin: 0 7.5px 15px;
  overflow: hidden;
  border: 6px solid rgba(255, 255, 255, 1);
  box-shadow: 1px -6px 20px 0px rgb(171 106 250 / 68%), 0 5px 15px #f9b1ba91;
  &:hover {
    cursor: pointer;
    & .background-image {
      transform: scale(1.1);
      transition: transform 6s cubic-bezier(0.25, 0.45, 0.45, 0.95);
    }
  }
  &:first-child {
    margin-right: 7.5px;
  }
  &:last-child {
    margin-left: 7.5px;
  }
  @media screen and (max-width: 800px) {
    flex-direction: column;
    height: 250px;
    margin: 0 7.5px 20px;
  }
`;

MenuItemContainer.displayName = "MenuItemContainer";

export const BackgroundImageContainer = styled.div`
  width: 100%;
  height: 100%;
  background-size: cover;
  background-position: center;
  background-image: ${({ imageUrl }) => `url(${imageUrl})`};
`;

BackgroundImageContainer.displayName = "BackgroundImageContainer";

export const ContentContainer = styled.div`
  height: 6em;
  width: 9em;
  padding: 0 25px;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  border: 3px solid rgba(255, 255, 255, 1);
  position: absolute;
  backdrop-filter: blur(3px);
  border-radius: 10px;
  span {
    font-weight: bold;
    color: transparent;
    background-image: linear-gradient(
      90deg,
      #ff6e7f 0%,
      #ff6e7f 50%,
      rgba(255, 255, 255, 1) 50%,
      rgba(255, 255, 255, 1) 100%
    );
    background-repeat: repeat;
    background-size: 200%;
    background-position: 100% 0;
    -webkit-background-clip: text;
    background-clip: text;
    transition: background-position 300ms;
  }
  &:hover span {
    background-position: 0 0;

    &::before {
      transform-origin: 0 0;
      transform: scale3d(1, 1, 1);
    }
  }
  &:hover {
    background: rgba(255, 255, 255, 1);
    transform: scale(1.25);
    transition: transform 0.6s cubic-bezier(0.25, 0.45, 0.45, 0.95),
      background 1s cubic-bezier(0.25, 0.45, 0.45, 0.95);
  }
  @media screen and (max-width: 800px) {
    width: 50%;
  }
  @media screen and (max-width: 300px) {
    width: 60%;
    &:hover {
    transform: scale(1.10);
  }
  }
`;

export const ContentTitle = styled.span`
  margin-bottom: 6px;
  font-size: 22px;
  font-weight: 700;
  @media screen and (max-width: 800px) {
    font-size: 28px;
  }
`;

export const ContentSubtitle = styled.span`
  font-size: 16px;

  @media screen and (max-width: 800px) {
    font-size: 20px;
  }
`;
