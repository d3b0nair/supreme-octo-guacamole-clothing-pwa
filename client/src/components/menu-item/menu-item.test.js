import React from "react";
import {
  render,
  fireEvent,
  waitFor,
  screen,
  cleanup,
} from "@testing-library/react";
import { MenuItem } from "./menu-item.component";

describe("MenuItem component", () => {
  const linkUrl = "/hats";
  const mockMatch = {
    url: "/shop",
  };

  const mockHistory = {
    push: jest.fn(),
  };
  const mockProps = (otherProps) => ({
    ...otherProps,
    match: mockMatch,
    history: mockHistory,
    linkUrl,
    size: "large",
    title: "hats",
    imageUrl: "testimage",
  });
  afterEach(cleanup);
  const component = (mockProps) => <MenuItem {...mockProps} />;

  it("should mount", () => {
    const { container } = render(component(mockProps()));
    expect(container).toMatchSnapshot();
  });

  it("should call history.push with the right string when MenuItemContainer clicked", async () => {
    render(component(mockProps()));
    const MenuItemContainer = screen.getByTestId("MenuItemContainer");
    await waitFor(() => fireEvent.click(MenuItemContainer));
    expect(mockHistory.push).toHaveBeenCalledWith(`${mockMatch.url}${linkUrl}`);
  });
});
