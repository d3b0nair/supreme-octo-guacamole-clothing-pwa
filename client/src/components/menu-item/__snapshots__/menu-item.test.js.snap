// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuItem component should mount 1`] = `
<div>
  <div
    class="sc-bdvvtL kaPJzZ"
    data-testid="MenuItemContainer"
  >
    <div
      class="sc-gsDKAQ fbXhpm background-image"
    />
    <div
      class="sc-dkPtRN deKDcd content"
    >
      <span
        class="sc-hKwDye YGhmR"
      >
        HATS
      </span>
      <span
        class="sc-eCImPb dBNPlT"
      >
        SHOP NOW
      </span>
    </div>
  </div>
</div>
`;
